### PR TITLE
Add exercise on variadic templates

### PR DESCRIPTION
**Pull request by bernhardmgruber**
Fixes: #263

### CODE REVIEW

File: code/variadic/README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+Feel free to comment out parts of main to test parts of your implementation sooner.
+All instructions are given in more detail in the source code as well.
+
+1. make `tuple` default construct
+2. avoid implicit construction from a single value
+3. write `head(tuple)`, which returns the first element of a tuple
+4. implement `get<I>(tuple)`
+5. specialize `std::tuple_size`
+6. specialize `std::tuple_element`
+7. implement `operator<<` for `tuple`
```
Comment (sponce) [inline code/variadic/README.md:16]:
```suggestion
7. implement simple `operator<<` for `tuple`
```

File: code/variadic/README.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+8. implement a binary `tuple_cat(tuple, tuple)`
+9. make `operator<<` produce a given output
```
Comment (sponce) [inline code/variadic/README.md:18]:
```suggestion
9. make complete `operator<<` producing comma separated output
```

File: code/variadic/solution/variadic.sol.cpp [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+#include <tuple>
+
+// we start with the tuple from the slides on variadic templates
+
+template <typename... Ts>
+struct tuple {};
+
+template <typename T, typename... Ts>
+struct tuple<T, Ts...> : tuple<Ts...> {
+    tuple() = default;
```
Comment (sponce) [inline code/variadic/solution/variadic.sol.cpp:14]:
```suggestion
    // exercise 1: activate the default constructor
    tuple() = default;
```

File: code/variadic/solution/variadic.sol.cpp [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+template <typename... Ts>
+auto& head(tuple<Ts...>& t) {
+    return t.m_head;
+}
+
+// exercise 4: implement get<I> here
+template <std::size_t I, typename... Ts>
+auto& get(tuple<Ts...>& t) {
+    if constexpr (I == 0)
+        return t.m_head;
```
Comment (sponce) [inline code/variadic/solution/variadic.sol.cpp:42]:
```suggestion
        return head(t);
```

**Comment by bernhardmgruber** [comment]
> I would simply drop exercise 7. The fold is not really trivial (and actually throws a wrong compiler warning to me that the statement has no effect ) and anyway the output is crap.

I think the fold `(os << ... << get<Is>(t));` is easier than `((os << get<Is>(t) << (Is == sizeof...(Ts) - 1 ? "" : ", ")), ...);`. And the first fold would be an example of a binary fold. Exercise 9 uses a unary fold over comma, which always feels like an abuse to me.

The reason you may get a compiler warning is because printing the empty tuple let's the fold collapse to just `os;`, which is indeed a statement without an effect.

**Comment by bernhardmgruber** [comment]
I could add another exercise where the students are supposed to implement a function `add(tuple<...> t)` that just does a fold over `+` and adds the tuple elements? That could be an example for a binary fold, since it should handle an empty tuple. What do you think?

**Comment by sponce** [comment]
> I could add another exercise where the students are supposed to implement a function `add(tuple<...> t)` that just does a fold over `+` and adds the tuple elements? That could be an example for a binary fold, since it should handle an empty tuple. What do you think?

sounds like a good idea indeed if you want to show the binary fold. At least I find it better than implementing a "bad" operator<< (in the sens that the output is unreadable without spaces or commas)

**Comment by bernhardmgruber** [comment]
I added an addition example to implement `auto sum(tuple<Ts...>)`.

File: code/variadic/README.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+and extend it step by step to make the example code in `main` compile.
+
+Feel free to comment out parts of main to test parts of your implementation sooner.
+All instructions are given in more detail in the source code as well.
+
+This example neglects const correctness so there is less code to write.
+Making it const correct would require adding several overloads with almost the same code as the non-const version.
+You can try this is a bonus exercise at the end!
+
+1. make `tuple` default construct
```
Comment (hageboeck) [inline code/variadic/README.md:14]: For me this doesn't parse.
- *a* default construct?
- default construct*able*
- Ensure that `tuple` has a default constructor

By looking at the solution, the best is presumably:
```suggestion
1. Ask the compiler to generate a default constructor for `tuple`
```

Comment (hageboeck) [reply code/variadic/README.md:14]: I would say that the instructions should be much more detailed. I struggle to understand what to do. More suggestions to follow soon

Comment (bernhardmgruber) [reply code/variadic/README.md:14]: And the statement means: "Make tuple<T...> t; compile". So, make `tuple` default constructible would be the best of your wordings.

Comment (bernhardmgruber) [reply code/variadic/README.md:14]: Yes, because this is where they should add some code as detailed inside `main` :)

Comment (hageboeck) [reply code/variadic/README.md:14]: Ah, OK. I didn't manage to read that far down before the meeting. Now continuing ...

Comment (bernhardmgruber) [reply code/variadic/README.md]: I rephrased `make tuple default construct` to `make tuple default-constructible` and pointed out more strongly in the README.md that the detailed instructions are to be found in `main`.

**Review by sponce — APPROVED** [review]

**Pull request merged**